cmd/pinniped-proxy/src/logging.rs

```diff
@@ -0,0 +1,108 @@
+use std::cmp::PartialEq;
+use std::fmt;
+
+use hyper::{Body, Request, Response};
+
+#[derive(Clone, Debug, Default, PartialEq)]
+/// LogData encapsulates all the data we want to collect during a proxied request.
+///
+/// This includes data from the initial request as well as the returned response.
+pub struct LogData {
+    target_uri: hyper::Uri,
+    method: hyper::Method,
+    // status_code is optional since we don't know it's value until the upstream
+    // response is returned.
+    status_code: Option<hyper::StatusCode>,
+}
+
+/// We implement the Display trait for our LogData struct so that it
+/// knows how to display itself as a string when requested.
+impl fmt::Display for LogData {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // If we have a status, use the string version of that,
+        // otherwise we'll just display a hyphen.
+        let status: String = match self.status_code {
+            Some(sc) => sc.to_string(),
+            None => "-".into(),
+        };
+        write!(f, "{} {} {}", self.method, self.target_uri, status)
+    }
+}
+
+/// request_log_data returns a new LogData with the relevant data from the request. 
+pub fn request_log_data(req: &Request<Body>) -> LogData {
+    LogData {
+        target_uri: req.uri().clone(),
+        method: req.method().clone(),
```

Comment:
Here's an example of cloning so that the LogData that we're returning does not refer to memory that is a part of the `req` reference - instead, we've created new copies of the data we need so that LogData is independent. If I don't clone them, the compiler tells us that we've implicitly moved the ownership of the uri from the `req` reference to the LogData struct, which violates concurrency principles and would stop the compiler from being able to determine when it can free the memory:
```
error[E0507]: cannot move out of a shared reference
  --> src/logging.rs:35:21
   |
35 |         target_uri: *req.uri(),
   |                     ^^^^^^^^^^ move occurs because value has type `Uri`, which does not implement the `Copy` trait

error: aborting due to previous error
```